### PR TITLE
Bugfix/myslq abstract table spec fixes

### PR DESCRIPTION
**Pull request by lummax**
This PR renames the MySQL support to MariaDB support and fixes the tests.

See https://github.com/outr/scalarelational/issues/11#issuecomment-129217160 for details.

### CODE REVIEW

**Comment by lummax** [comment]
It seems travis-ci somehow got confused by my heavy use of `git rebase` :/

It does not use MariaDB as stated in the travis-ci config.

File: README.md [hunk 1]
```diff
@@ -10,9 +10,9 @@ making working with your database easier, but this comes at the cost of ineffici
 confusing code mapping. The goal of ScalaRelational is to provide a one-to-one representation of the SQL language in
 Scala and provide performance improvements to working with the database.
 
-Currently H2 and MySQL is tested, but other databases technically are supported though certain functionality may require
-some slight alterations to work properly. One of the goals of ScalaRelational is to be a straight-forward, type-safe framework
-to write SQL queries in Scala with no magic or guesswork of what the resulting SQL will be.
+Currently H2 and MariaDB (version 10.0) is tested, but other databases technically are supported though certain functionality
```
Comment (darkfrog26) [inline README.md]: Since MySQL and MariaDB should be almost entirely compatible we should probably list MySQL here as well.

File: core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala [hunk 2]
```diff
@@ -12,9 +13,9 @@ import org.scalarelational.model.Datastore
  * @author Matt Hicks <matt@outr.com>
  */
 class ObjectSerializationConverter[T <: AnyRef] extends DataType[T] {
-  override def jdbcType = Types.BINARY
+  override def jdbcType = Types.BLOB
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala]: Why did you change this to `BLOB`?

File: core/src/test/scala/org/scalarelational/AbstractTableSpec.scala [hunk 3]
```diff
@@ -546,30 +546,30 @@ trait AbstractTableSpec extends WordSpec with Matchers {
...
     val date = column[Option[Timestamp]]("date")
   }
   object suppliers extends Table("SUPPLIER") {
     val id = column[Option[Int]]("SUP_ID", PrimaryKey, AutoIncrement)
-    val name = column[String]("SUP_NAME")
-    val street = column[String]("STREET")
-    val city = column[String]("CITY")
-    val state = column[String]("STATE")
-    val zip = column[String]("ZIP")
+    val name = column[String]("SUP_NAME", ColumnLength(200))
```
Comment (darkfrog26) [inline core/src/test/scala/org/scalarelational/AbstractTableSpec.scala]: We should be able to rely on defaults. If we can't then something is wrong with the implementation.

File: mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala [hunk 4]
```diff
@@ -43,6 +48,12 @@ abstract class MySQLDatastore private() extends SQLDatastore with Logging {
     }
   }
 
+  override def ddl(drop: DropTable): List[CallableInstruction] = List(
+      CallableInstruction("SET foreign_key_checks = 0;"),
```
Comment (darkfrog26) [inline mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala]: This should only be set if `cascade` is true.

File: README.md (hunk 1 above)
Comment (lummax) [inline README.md]: Maybe. But as the travis tests only test MariaDB support we can't say for sure if MySQL is supported (and to what extend). Especially for in the future. We could add a setence to the README about the problems we know of.

File: core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala (hunk 2 above)
Comment (lummax) [inline core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala]: Because MariaDB has a maximum row size. A `BINARY` column adds to the used row size. A `BLOB` does not. It works for H2 as they have no limit, but with MariaDB we could not serialize abitrary objects and add them to a row. A consequence would be, that in some cases it works (sum of all objects < row_size_limit) and in some it misteriously does not.

File: mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala (hunk 4 above)
Comment (lummax) [inline mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala]: Right! I'll change it.

File: core/src/test/scala/org/scalarelational/AbstractTableSpec.scala (hunk 3 above)
Comment (lummax) [inline core/src/test/scala/org/scalarelational/AbstractTableSpec.scala]: Then we would need some magic to automatically determine a `ColumnLength` for MariaDB. An an additional check if the sum of the columns (and every other datatype that adds to the row length) exceeds the limit.

This change fixes the tests. And I think we could expect users of scalarelational that want to use MariaDB to know about the row size limitation and introduce a `ColumnLength` in their code.

Setting `DefaultVarCharLength` for MariaDB to say `200` might be an option if we document this, including the implications (set `ColumnLength` if your string is expected to exceed 200 chars, set `ColumnLength` if you use to many `String` columns, ..)

File: README.md (hunk 1 above)
Comment (darkfrog26) [inline README.md]: If I have MySQL installed instead of MariaDB wouldn't the tests be run against it instead?

File: core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala (hunk 2 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala]: Ah, that's a good fix.

File: core/src/test/scala/org/scalarelational/AbstractTableSpec.scala (hunk 3 above)
Comment (darkfrog26) [inline core/src/test/scala/org/scalarelational/AbstractTableSpec.scala]: I think that's a better solution.

File: README.md (hunk 1 above)
Comment (lummax) [inline README.md]: Sure, as MariaDB is a "drop-in" replacement for MySQL. But the current travis test use the `mariadb` plugin (https://github.com/lummax/scalarelational/commit/64c53dbabc8ba3eb195be90470ff639674439135) so the automated tests don't check for MySQL compliance.

Comment (darkfrog26) [inline README.md]: I believe I read that the MariaDB plugin works for MySQL and the MySQL plugin works for MariaDB. Is that not the case?

File: core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala [hunk 5]
```diff
@@ -36,6 +37,17 @@ class ObjectSerializationConverter[T <: AnyRef] extends DataType[T] {
...
+      try {
+        ois.readObject().asInstanceOf[T]
+      } finally {
+        ois.close()
+      }
+    }
+    // The mysql-java-connector does not return a Blob, but a byte array. This is a violation
+    //of the JDBC 4.1 specification (appendix B: `Data Conversion Tables`).
+    // http://dev.mysql.com/doc/connector-j/en/connector-j-reference-type-conversions.html
     case array: Array[Byte] => {
```
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala]: Is this ``case`` block still needed?

File: mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
+package org.scalarelational.mariadb
+
+;
```
Comment (tindzk) [inline mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala]: Remove semicolon

File: mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  override def ddl(drop: DropTable): List[CallableInstruction] =
+    if (drop.cascade) List(CallableInstruction("SET foreign_key_checks = 0;"),
+                           super.ddl(drop).head,
+                           CallableInstruction("SET foreign_key_checks = 1;"))
+    else super.ddl(drop)
+
+  def updateDataSource() = {
+    dispose() // Make sure to shut down the previous DataSource if possible
+    val source: MysqlDataSource = new MysqlDataSource()
+    source.setURL("jdbc:mysql://" + config().host + "/" + config().schema)
```
Comment (tindzk) [inline mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala]: This should be ``jdbc:mariadb://...``.

File: README.md (hunk 1 above)
Comment (lummax) [inline README.md]: Ok, I think we might misunderstand eachother:

- You can use MariaDB as a drop-in replacement for MySQL. And I believe the other way arround is possible.
- Some features missing from MySQL are implemented in MariaDB. So if we test against MariaDB me might miss regressions for MySQL.
- Travis has a pre-installed MySQL instance running. But can be configured (and is using this PR) to use MariaDB instead. I don't believe is easy to enable both (of course it would be writing custom install scripts).

Comment (darkfrog26) [inline README.md]: I would expect we would be writing for the lowest common denominator so that we can support both. Is there anything in MariaDB that we're using that's not available in MySQL?

Comment (lummax) [inline README.md]: So far only, that "IF NOT EXISTS" for "CREATE INDEX" is not supported on MySQL.

I would add a comment about that to the README•

File: mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala (hunk 7 above)
Comment (lummax) [inline mariadb/src/main/scala/org/scalarelational/mariadb/MariaDBDatastore.scala]: No.

It tried switching to the mariadb-jdbc package, but some things are different and all the tests failed again. And if we would switch we might loose compatibility back to MySQL.

File: core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala (hunk 5 above)
Comment (lummax) [inline core/src/main/scala/org/scalarelational/datatype/ObjectSerializationConverter.scala]: Yes, as we use the bugged Mysql-jdbc Driver that returnes an `Array[Byte]`.

**Comment by lummax** [comment]
I don't get why travis keeps failing. Travis works and does not fail on my branch: https://travis-ci.org/lummax/scalarelational/builds/75288134

**Comment by darkfrog26** [comment]
The most likely scenario is that they are running an older version of MariaDB.

Looks like `IF NOT EXISTS` isn't supported on indexes in that version of the database.

**Comment by lummax** [comment]
It's weird. I have the line `Installing MariaDB version 10.0` in my travis log. This line does not appear in the travis log for this PR.

**Comment by darkfrog26** [comment]
@lummax any updates on this? I just merged a pretty major update to master that will additional flexibility but unfortunately will also require a few code changes.

**Comment by lummax** [comment]
I will update it tomorrow!

**Comment by darkfrog26** [comment]
@lummax great, thanks for the update.
